src/VisualStudio/Core/Def/DocumentOutline/DocumentSymbolDataViewModelSorter.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Windows.Data;
+using System.Windows.Markup;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Sorts immutable collections of <see cref="DocumentSymbolDataViewModel"/>s 
+    /// </summary>
+    internal class DocumentSymbolDataViewModelSorter : MarkupExtension, IMultiValueConverter
```

Comment:
what does MarkupExtension mean?

Reply:
`MarkupExtension` is essentially a factory pattern that tells WPF how to construct a type. The `ProvideValue` method is what WPF calls when it needs to construct an instance of this type. Without this we would not be able to refer to an instance of this type in XAML as we do in `DocumentOutlineView.xaml`

```xaml
<!-- An instance of our custom sorter, This uses ImmutableArray's sort methods instead of linq 
     which uses far less memory.
-->
<self:DocumentSymbolDataViewModelSorter x:Key="ItemSorter"/>
```

Reply:
sealed.